### PR TITLE
[FLINK-23663][table-planner] Reduce state size of ChangelogNormalize

### PR DESCRIPTION
**Pull request by Airblader**
## What is the purpose of the change

`ChangelogNormalize` keeps state for each unique key, which is quite expensive. In general it isn't safe to push filters through this operator, but we _can_ allow filters which only filter on the key space to go past it. This allows for a significant reduction in the footprint of the operator if the key space is reduced.

This PR implements a new optimization rule which pushes such primary key filters past the `ChangelogNormalize` operator to allow its state to be reduced. The tests have first been run without the rule and the changes made by the rule are pushed as a second fixup commit for a more pleasant review experience. :-)

Note that sources which implement `SupportsFilterPushDown` currently always have all filters pushed past `ChangelogNormalize` into the source, see FLINK-24014.

~This PR is currently in draft mode as we want to implement FLINK-24004 here to improve the test infrastructure.~

## Verifying this change

This change added tests and can be verified as follows:

* `PushFilterPastChangelogNormalizeRuleTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+ * operator to reduce its state size.
+ *
+ * <p>This rule looks for Calc → ChangelogNormalize where the {@link StreamPhysicalCalc Calc}
+ * contains a filter condition. The condition is transformed into CNF and then each conjunction is
+ * tested for whether it affects only primary key columns. If such conditions exist, they are moved
+ * into a new, separate Calc and pushed through the ChangelogNormalize operator. ChangelogNormalize
+ * keeps state for every unique key it encounters, thus pushing filters on the primary key in front
+ * of it helps reduce the size of its state.
+ *
+ * <p>Note that pushing primary key filters is safe to do, but pushing any other filters would lead
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java:59]: Wonky sentence here.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * public class CustomSourceTest {
+ *     {@literal @}Rule public SharedObjects sharedObjects = SharedObjects.create();
+ *
+ *     {@literal @}Test
+ *     public void test() {
+ *         SharedReference<List<Long>> appliedLimits = sharedObjects.add(new ArrayList<>());
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java:76]: Actually using SharedObjects here is entirely unnecessary. We instantiate the source specifically for this test, so we can just keep state on the source and assert on that. Will update and also the docs for the harness.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        }
+
+        @Override
+        public Set<ConfigOption<?>> requiredOptions() {
+            return new HashSet<>();
+        }
+
+        @Override
+        public Set<ConfigOption<?>> optionalOptions() {
+            return Collections.singleton(SOURCE);
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java:156]: Missing SINK

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+     * Serializable base class for custom sources which implements {@link ScanTableSource}.
+     *
+     * <p>Most interface methods are default-implemented for convenience, but can be overridden when
+     * necessary. By default, a {@link ScanRuntimeProvider} is used which doesn't produce anything.
+     *
+     * <p>Sources derived from this base class will also be provided the {@link
+     * DynamicTableFactory.Context} of the factory which gives access to e.g. the {@link
+     * CatalogTable}.
+     */
+    public abstract static class SourceBase implements Serializable, ScanTableSource {
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java:261]: Maybe rename to ScanSourceBase and also have a LookupSourceBase?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+ * to incorrect results.
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java:60]: nit: `potentially incorrect results` it depends on the query

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+     * anything else.
+     */
+    private void partitionPrimaryKeyPredicates(
+            List<RexNode> predicates,
+            Set<Integer> primaryKeyIndices,
+            List<RexNode> primaryKeyPredicates,
+            List<RexNode> remainingPredicates) {
+        for (RexNode predicate : predicates) {
+            int[] inputRefs = extractRefInputFields(Collections.singletonList(predicate));
+            if (Arrays.stream(inputRefs).allMatch(primaryKeyIndices::contains)) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java:115]: I'm just mentioning here if it was a conscious decision to support use cases such as:
primary key with `(a,b,c)` with a predicate `a=1 && b=42` would be pushed through?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                primaryKeyPredicates.add(predicate);
+            } else {
+                remainingPredicates.add(predicate);
+            }
+        }
+    }
+
+    /** Pushes {@param primaryKeyPredicates} into the {@link StreamPhysicalChangelogNormalize}. */
+    private StreamPhysicalChangelogNormalize pushFiltersThroughChangeNormalize(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java:124]: nit: `pushFiltersThroughChangelogNormalize`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+                    calc.getProgram().expandLocalRef(calc.getProgram().getCondition()));
+        }
+
+        for (Pair<RexLocalRef, String> projectRef :
+                projectFromCalc.getProgram().getNamedProjects()) {
+            final RexNode project = projectFromCalc.getProgram().expandLocalRef(projectRef.left);
+            programBuilder.addProject(project, projectRef.right);
+        }
+
+        final RexProgram newProgram = programBuilder.getProgram();
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRule.java:187]: will the resulting program have local refs again? we are not using them yet but just for the future, we should restore a compact representation after applying this rule

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRuleTest.xml [hunk 9]
```diff
@@ -30,10 +30,11 @@ LogicalAggregate(group=[{0}], sum=[SUM($0)])
     </Resource>
     <Resource name="optimized rel plan">
       <![CDATA[
-Calc(select=[f1, f1 AS sum], where=[AND(<(f1, 10), OR(>(f1, 3), IS NULL(f2)))])
+Calc(select=[f1, f1 AS sum], where=[OR(>(f1, 3), IS NULL(f2))])
 +- ChangelogNormalize(key=[f1])
-   +- Exchange(distribution=[hash[f1]])
-      +- TableSourceScan(table=[[default_catalog, default_database, T]], fields=[f0, f1, f2])
+   +- Calc(select=[f0, f1, f2], where=[<(f1, 10)])
+      +- Exchange(distribution=[hash[f1]])
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/rules/physical/stream/PushFilterPastChangelogNormalizeRuleTest.xml:36]: sorry, if I was not clear enough. but of course we can also push it through the exchange that will (always?) be in front of the `ChangelogNormalize`. Shuffling is also expensive due to network IO, the more filtering here, the better.

**Review by twalthr — COMMENTED** [review]
Thanks @Airblader. Very good code quality. It's a pleasure to read your PRs :) I added some comments.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java [hunk 10]
```diff
@@ -108,39 +101,24 @@
...
-                .option(SINK, sink.serialize())
-                .build();
+    /**
+     * Creates a builder for a new {@link TableDescriptor} specialized for this harness.
+     *
+     * <p>Use this method to create a {@link TableDescriptor} and passing in the source / sink
+     * implementation you want to use. The descriptor can for example be used with {@link
+     * TableEnvironment#createTable(String, TableDescriptor)} to register a table.
+     */
+    public static HarnessTableDescriptor.Builder forHarness() {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java:113]: nit: `newBuilder()`? `TableFactoryHarness.forHarness` reads cumbersome

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableDescriptor.java [hunk 11]
```diff
@@ -51,14 +51,14 @@
...
-public final class TableDescriptor {
+public class TableDescriptor {
 
     private final @Nullable Schema schema;
     private final Map<String, String> options;
     private final List<String> partitionKeys;
     private final @Nullable String comment;
 
-    private TableDescriptor(
+    public TableDescriptor(
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableDescriptor.java:61]: `protected`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DefaultCatalogTable.java [hunk 12]
```diff
@@ -33,14 +33,14 @@
...
-class DefaultCatalogTable implements CatalogTable {
+public class DefaultCatalogTable implements CatalogTable {
 
     private final Schema schema;
     private final @Nullable String comment;
     private final List<String> partitionKeys;
     private final Map<String, String> options;
 
-    DefaultCatalogTable(
+    public DefaultCatalogTable(
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DefaultCatalogTable.java:43]: `protected`

**Review by twalthr — COMMENTED** [review]
Changes look good to me. I had some last visibility suggestions.

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request closed without merging**